chainstate/tx-verifier/src/transaction_verifier/mod.rs

```diff
@@ -589,48 +688,224 @@ where
+                    }
+                },
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+
+        let output_undos = tx
+            .outputs()
+            .iter()
+            .filter_map(|output| match output {
+                TxOutput::Transfer(_, _)
+                | TxOutput::Burn(_)
+                | TxOutput::CreateStakePool(_, _)
+                | TxOutput::ProduceBlockFromStake(_, _)
+                | TxOutput::CreateDelegationId(_, _)
+                | TxOutput::DelegateStaking(_, _)
+                | TxOutput::LockThenTransfer(_, _, _)
+                | TxOutput::IssueNft(_, _, _) => None,
+                TxOutput::IssueFungibleToken(issuance_data) => {
+                    let result = make_token_id(tx.inputs())
+                        .ok_or(ConnectTransactionError::TokensError(
+                            TokensError::TokenIdCantBeCalculated,
+                        ))
+                        .and_then(
+                            |token_id| -> Result<tokens_accounting::TokenAccountingUndo, _> {
+                                let data = tokens_accounting::TokenData::FungibleToken(
+                                    issuance_data.as_ref().clone().into(),
+                                );
+                                self.tokens_accounting_cache
+                                    .issue_token(token_id, data)
+                                    .map_err(ConnectTransactionError::TokensAccountingError)
+                            },
+                        );
+                    Some(result)
+                }
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+
+        // Store accounting operations undos
+        if !input_undos.is_empty() || !output_undos.is_empty() {
+            let tx_undos = input_undos.into_iter().chain(output_undos).collect();
+            self.tokens_accounting_block_undo
+                .get_or_create_block_undo(&tx_source.into())
+                .insert_tx_undo(tx.get_id(), tokens_accounting::TxUndo::new(tx_undos))
+                .map_err(ConnectTransactionError::TokensAccountingBlockUndoError)
+        } else {
+            Ok(())
+        }
+    }
+
+    fn disconnect_tokens_accounting_outputs(
+        &mut self,
+        tx_source: TransactionSource,
+        tx: &Transaction,
+    ) -> Result<(), ConnectTransactionError> {
+        // apply undos to accounting
+        let block_undo_fetcher = |id: Id<Block>| {
+            self.storage
+                .get_tokens_accounting_undo(id)
+                .map_err(|_| ConnectTransactionError::TxVerifierStorage)
+        };
+        let undos = self.tokens_accounting_block_undo.take_tx_undo(
+            &tx_source,
+            &tx.get_id(),
+            block_undo_fetcher,
+        )?;
+        if let Some(undos) = undos {
+            undos
+                .into_inner()
+                .into_iter()
+                .try_for_each(|undo| self.tokens_accounting_cache.undo(undo))?;
```

Comment:
it might mess up with DeltaData, I'll check

Reply:
An interesting case that worked anyway because of the lack of checks on undo is mint + lock in the same tx, which should've failed with an error but worked anyway. I added additional checks and reversed the order of processing the undo.

Also I reversed it for pos accounting outputs, but I can't see the case to reproduce